packages/commonwealth/server/controllers/server_groups_methods/update_group.ts

```diff
@@ -177,3 +154,73 @@ export async function __updateGroup(
 
   return [group.toJSON(), analyticsOptions];
 }
+
+async function updateGroupPermissions(
+  topics: number[],
+  allowed_actions: ForumActions[],
+  group_id: number,
+  models: DB,
+  transaction: Transaction,
+): Promise<GroupPermissionInstance[]> {
+  const existingGroupPermissions = await models.GroupPermission.findAll({
+    where: { group_id },
+  });
+
+  const existingTopicIds = existingGroupPermissions.map(
+    (permission) => permission.topic_id,
+  );
+
+  const permissionsToRemove = existingTopicIds.filter(
+    (id) => !topics.includes(id!),
+  );
+  const permissionsToUpsert = topics.filter(
+    (id) => !permissionsToRemove.includes(id!),
+  );
+
+  if (permissionsToRemove.length > 0) {
+    await models.GroupPermission.destroy({
+      where: {
+        group_id,
+        topic_id: { [Op.in]: permissionsToRemove },
+      } as WhereOptions<GroupPermissionAttributes>,
+      transaction,
+    });
+  }
+
+  if (permissionsToUpsert.length === 0) {
+    return Promise.resolve([]);
+  }
+
+  const upsertGroupPermissions = permissionsToUpsert.map((topic_id) => ({
+    group_id,
+    topic_id,
+    allowed_actions,
+  }));
+
+  const upsertPromises = upsertGroupPermissions.map((p) =>
+    models.sequelize.query(
+      `
+            INSERT INTO "GroupPermissions" (group_id, topic_id, allowed_actions, created_at, updated_at) VALUES
+            (:groupId, :topicId, Array[:allowedActions]::"enum_GroupPermissions_allowed_actions"[], NOW(), NOW())
+            ON CONFLICT (group_id, topic_id) DO UPDATE
+            SET
+                allowed_actions = EXCLUDED.allowed_actions,
+                updated_at = EXCLUDED.updated_at
+            RETURNING *;
```

Comment:
We should always use a type for raw queries, e.g. `models.sequelize.query<GroupPermissionInstance>(...)`, then there's no need to cast the type later.

Also, I think wildcards `*` should be avoided. It's easier to write but tricker to debug due to lack of transparency.